src/Java.Runtime.Environment/Java.Interop/ManagedValueManager.cs

```diff
@@ -0,0 +1,237 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Java.Interop {
+
+	class ManagedValueManager : JniRuntime.JniValueManager {
+
+		Dictionary<int, List<IJavaPeerable>>?   RegisteredInstances = new Dictionary<int, List<IJavaPeerable>>();
+
+		public override void WaitForGCBridgeProcessing ()
+		{
+		}
+
+		public override void CollectPeers ()
+		{
+			if (RegisteredInstances == null)
+				throw new ObjectDisposedException (nameof (MonoRuntimeValueManager));
+
+			var peers = new List<IJavaPeerable> ();
+
+			lock (RegisteredInstances) {
+				foreach (var ps in RegisteredInstances.Values) {
```

Comment:
Here it might be more efficient to do:
```
                       List<IDisposable> values;
                       lock (RegisteredInstances) {
                               values = new List<IDisposable> (RegisteredInstances.Values);
                               RegisteredInstances.Clear ();
                        }
```
and do the rest of cleaning after that? It will also lower time spent in lock.

Reply:
`RegisteredInstances.Values` is of type `ICollection<List<IJavaPeerable>>`, not `ICollection<IJavaPeerable>`.  Thus, `new List<IDisposable> (RegisteredInstances.Values)` won't work.

Reply:
I was updating the code from elswhere, so it would need to be `new List<List<IJavaPeerable>> (RegisteredInstances.Values)`, the idea is the same though. To leave the `lock` block quickly.